File: Inversion.php
<?php

// [5,3,4,2,1]
// [5,4,3,2,1]
// 0 1 2 3 4
// function inversion(array $array) {
// 	$result = 0;
// 	$count = count($array); //5

// 	$data = [];
// 	for ($i = 0; $i < $count - 2; $i++) {
//         for ($j = $i + 1; $j < $count - 1; $j++) {
//             for ($k = $j + 1; $k < $count; $k++) {
//                 if ($array[$i] > $array[$j] && $array[$j] > $array[$k]) {
//                     $result++;
//                 }
//             }
//         }
//     }
// 	return $result;
// }

// $array = [5,3,4,2,1];
// $result = inversion($array);
// echo $result;

function countInversions(array &$arr, $temp, $left, $right) { // left = 0 right = 4
    $inversions = 0;

    if ($left < $right) {
        $mid = (int)(($left + $right) / 2);

        $inversions += countInversions($arr, $temp, $left, $mid);
        $inversions += countInversions($arr, $temp, $mid + 1, $right);
        $inversions += merge($arr, $temp, $left, $mid, $right);
    }

    return $inversions;
}

function merge(array &$arr, &$temp, $left, $mid, $right) {
    $i = $left;
    $j = $mid + 1;
    $k = $left;
    $inversions = 0;

    while ($i <= $mid && $j <= $right) {
        if ($arr[$i] <= $arr[$j]) {
            $temp[$k] = $arr[$i];
            $i++;
        } else {
            $temp[$k] = $arr[$j];
            $inversions += ($mid - $i + 1); // Count inversions
            $j++;
        }
        $k++;
    }

    while ($i <= $mid) {
        $temp[$k] = $arr[$i];
        $i++;
        $k++;
    }

    while ($j <= $right) {
        $temp[$k] = $arr[$j];
        $j++;
        $k++;
    }

    for ($i = $left; $i <= $right; $i++) {
        $arr[$i] = $temp[$i];
    }

    return $inversions;
}

function inversion(array $array) {
    $temp = $array;
    $count = count($array);
    $result = countInversions($array, $temp, 0, $count - 1);

    return $result;
}

// $array = [5,3,4,2,1];
$array = [4,1,3,2,5];
$result = inversion($array);
echo $result;

File: algorithm-go/Inversion.go
package main

import (
	"fmt"
)

func main() {
	arr := []int{5, 3, 4, 2, 1}

	inversions, inversionList := countInversions(arr)

	fmt.Println("Number of inversions:", inversions)
	fmt.Println("Individual inversions:")
	for _, inv := range inversionList {
		fmt.Println(inv)
	}
}

func countInversions(arr []int) (int, [][]int) {
	_, inversions, inversionList := mergeSort(arr)
	return inversions, inversionList
}

func mergeSort(arr []int) ([]int, int, [][]int) {
	count := len(arr)

	if count <= 1 {
		return arr, 0, [][]int{} // Return empty inversion list for base case
	}

	mid := count / 2
	left, leftInversions, leftList := mergeSort(arr[:mid])
	right, rightInversions, rightList := mergeSort(arr[mid:])

	merged, splitInversions, splitInversionList := merge(left, right)

	totalInversions := leftInversions + rightInversions + splitInversions

	// Concatenate inversion lists correctly
	inversionList := append(leftList, rightList...)
	inversionList = append(inversionList, splitInversionList...)

	return merged, totalInversions, inversionList
}

func merge(left, right []int) ([]int, int, [][]int) {
	merged := make([]int, 0, len(left)+len(right))
	inversions := 0
	i, j := 0, 0

	inversionList := [][]int{} // Initialize inversion list

	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			merged = append(merged, left[i])
			i++
		} else {
			inversions += len(left) - i
			merged = append(merged, right[j])
			j++
			inversionList = append(inversionList, []int{left[i], right[j]}) // Record inversion pair
		}
	}

	merged = append(merged, left[i:]...)
	merged = append(merged, right[j:]...)

	return merged, inversions, inversionList
}
